packages/rest/__tests__/session/CredsForSessCfg.test.ts

```diff
@@ -0,0 +1,211 @@
+            intialSessCfg, args
+        );
+        expect(sessCfgWithCreds.hostname).toBe("SomeHost");
+        expect(sessCfgWithCreds.tokenValue).toBe("FakeToken");
+        expect(sessCfgWithCreds.type).toBe(SessConstants.AUTH_TYPE_TOKEN);
+        expect(sessCfgWithCreds.tokenType).toBe(SessConstants.TOKEN_TYPE_APIML);
+        expect(sessCfgWithCreds.user).toBeUndefined();
+        expect(sessCfgWithCreds.password).toBeUndefined();
+    });
+
+    it("not prompt when asked not to prompt", async() => {
+        const intialSessCfg = {
+            hostname: "SomeHost",
+            port: 11,
+            rejectUnauthorized: true,
+        };
+        const args = {
+        };
+
+        let sessCfgWithCreds: ISession;
+        sessCfgWithCreds = await CredsForSessCfg.addCredsOrPrompt<ISession>(
+            intialSessCfg, args, {doPrompting: false}
+        );
+        expect(sessCfgWithCreds.type).toBe(SessConstants.AUTH_TYPE_BASIC);
+        expect(sessCfgWithCreds.user).toBeUndefined();
+        expect(sessCfgWithCreds.password).toBeUndefined();
+        expect(sessCfgWithCreds.tokenType).toBeUndefined();
+        expect(sessCfgWithCreds.tokenValue).toBeUndefined();
+    });
+
+    it("timeout waiting for user name", async() => {
+        const intialSessCfg = {
+            hostname: "SomeHost",
+            port: 11,
+            rejectUnauthorized: true,
+        };
+        const args = {
+            password: "FakePassword",
+        };
+
+        /* Jest hangs forever when you request data from stdin.
+         * mock-stdin should let you supply data. We could not make that work.
+         * At least with mock-stdin we only get our expected 30 second timout.
+         */
+        const mockStdIn = require("mock-stdin").stdin();
+        mockStdIn.send("FakeUser\n");   // this does not work
+        mockStdIn.end();
+
+        /* jest redirects stdin and we could not override it.
+         * So we catch the timeout error.
+         */
+        let sessCfgWithCreds: ISession;
+        let caughtError;
+        try {
+            sessCfgWithCreds = await CredsForSessCfg.addCredsOrPrompt<ISession>(
+                intialSessCfg, args
+            );
+        } catch (thrownError) {
+            caughtError = thrownError;
+        }
+        expect(caughtError instanceof ImperativeError).toBe(true);
+        expect(caughtError.message).toBe("We timed-out waiting for user name.");
+    }, waitForPrompt);
+
+    it("timeout waiting for password", async() => {
+        const intialSessCfg = {
+            hostname: "SomeHost",
+            port: 11,
+            rejectUnauthorized: true,
+        };
+        const args = {
+            user: "FakeUser",
+        };
+
+        /* Jest hangs forever when you request data from stdin.
+         * mock-stdin should let you supply data. We could not make that work.
+         * At least with mock-stdin we only get our expected 30 second timout.
```

Comment:
Wondering if it'd be possible to use the `jest.advanceTimerByTime(ms)` function
https://jestjs.io/docs/en/timer-mocks#advance-timers-by-time

We may prevent this single file from taking over a minute to run. 😋

Reply:
jest.advanceTimersByTime is meant to be run after an asynchronous operation has been started. However, we await our prompt for credentials making it synchronous. Calling advanceTimersByTime after the prompt is too late. I even tried calling advanceTimersByTime  before and after the prompt. The test hung until it hit a jest global timeout of 3 minutes.

Because you made me look at this problem again, I did find an alternative approach. I mock our CliUtils.sleep function:

    const sleepReal = CliUtils.sleep;
    CliUtils.sleep = jest.fn();

Now the entire test suite runs in about 1.5 seconds. Thanks for prodding me.